src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,305 @@
+
+        :param exchange_name: Name of the topic exchange
+        :type exchange_name: str
+        :param connection_url: Kombu connection URL
+        :type connection_url: str
+        """
+        queue_kwargs = [
+            {'routing_key': 'jobs.clusterexec'}
+        ]
+        super().__init__(exchange_name, connection_url, queue_kwargs)
+
+    def _execute(self, model_instance, command_args, finished_hosts_key):
+        """
+        Remotely executes OS-specific shell commands across a cluster.
+
+        :param model_instance: Initial model for the async operation
+        :type model_instance: commissaire.models.Model
+        :param command_args: Command name + arguments as a tuple
+        :type command_args: tuple
+        :param finished_hosts_key: Model attribute name for finished hosts
+        :type finished_hosts_key: str
+        """
+        # Split out the command name.
+        command_name = command_args[0]
+        command_args = command_args[1:]
+
+        end_status = 'finished'
+
+        # XXX We assume the model instance names a cluster.
+        #     Note, cluster_name is used in the except clause,
+        #     so it must be reliably defined.
+        cluster_name = getattr(model_instance, 'name', None)
+
+        try:
+            # Set the initial status in the store.
+            assert cluster_name is not None
+            model_json_data = model_instance.to_json()
+            self.logger.info('Setting initial status.')
+            self.logger.debug('Status={}'.format(model_json_data))
+            params = {
+                'model_type_name': model_instance.__class__.__name__,
+                'model_json_data': model_json_data
+            }
+            self.request('storage.save', params=params)
+        except Exception as error:
+            self.logger.error(
+                'Unable to save initial state for "{}" clusterexec due to '
+                '{}: {}'.format(cluster_name, type(error), error))
+            raise error
+
+        # Collect all host addresses in the cluster.
+
+        try:
+            cluster = Cluster.new(name=cluster_name)
+            params = {
+                'model_type_name': cluster.__class__.__name__,
+                'model_json_data': cluster.to_json(),
+                'secure': True
+            }
+            response = self.request('storage.get', params=params)
+            cluster = Cluster.new(**response['result'])
+        except Exception as error:
+            self.logger.warn(
+                'Unable to continue for cluster "{}" due to '
+                '{}: {}'.format(cluster_name, type(error), error))
+            raise error
+
+        n_hosts = len(cluster.hostset)
+        if n_hosts:
+            self.logger.debug(
+                '{} hosts in cluster "{}"'.format(n_hosts, cluster_name))
+        else:
+            self.logger.warn('No hosts in cluster "{}"'.format(cluster_name))
+
+        # TODO Find a better way to do this...
```

Comment:
@gbraad good point. This is a hold over from MVP but we should make a task to fix it.